SpeechService/Translations.cs

```diff
@@ -748,109 +757,159 @@ public static string Humanize(decimal? value)
                 order = "";
                 nextDigit = (int)((value - number) * 10);
             }
+            else if (digits < 3)
+            {
+                // Hundreds
+                (number, nextDigit) = Normalize(value, 100);
+                order = Properties.Phrases.hundred;
+            }
             else if (digits < 6)
             {
                 // Thousands
-                number = (int)(value / 1E3M);
-                order = " " + Properties.Phrases.thousand;
-                nextDigit = (int)((value - (number * 1E3M)) / 1E2M);
+                (number, nextDigit) = Normalize(value, 1E3M);
+                order = Properties.Phrases.thousand;
             }
             else if (digits < 9)
             {
                 // Millions
-                number = (int)(value / 1E6M);
-                order = " " + Properties.Phrases.million;
-                nextDigit = (int)((value - (number * 1E6M)) / 1E5M);
+                (number, nextDigit) = Normalize(value, 1E6M);
+                order = Properties.Phrases.million;
             }
             else if (digits < 12)
             {
                 // Billions
-                number = (int)(value / 1E9M);
-                order = " " + Properties.Phrases.billion;
-                nextDigit = (int)((value - (number * 1E9M)) / 1E8M);
+                (number, nextDigit) = Normalize(value, 1E9M);
+                order = Properties.Phrases.billion;
             }
             else if (digits < 15)
             {
                 // Trillions
-                number = (int)(value / 1E12M);
-                order = " " + Properties.Phrases.trillion;
-                nextDigit = (int)((value - (number * 1E12M)) / 1E11M);
+                (number, nextDigit) = Normalize(value, 1E12M);
+                order = Properties.Phrases.trillion;
             }
             else
             {
                 // Quadrillions
-                number = (int)(value / 1E15M);
-                order = " " + Properties.Phrases.quadrillion;
-                nextDigit = (int)((value - (number * 1E15M)) / 1E14M);
+                (number, nextDigit) = Normalize(value, 1E15M);
+                order = Properties.Phrases.quadrillion;
             }
 
-            // See if we have an exact match
-            if (((long)(((decimal)value) / (decimal)Math.Pow(10, digits - 1))) * (decimal)(Math.Pow(10, digits - 1)) == value)
+            // emit exact value
+            if (order == "")
             {
-                return minus + number + order;
+                // we recurse only if we know that there _is_ a nested order (i. e. stuff like "51 hundred thousand")
+                // otherwise we won't be able to round correctly
+                Debug.Assert(recursed == false);
+                return maybeMinus + number;
             }
 
-            // Describe decimal values
-            if (number < 100)
+            // recurse to handle nested orders (i. e. "50 hundred thousand")
```

Comment:
Do you mean 5 hundred thousand?

Reply:
Does not matter. A "nested order" is, for example, "hundred thousand".